### DISCUSSION TITLE
Option to find objects without throwing ObjectNotFoundError

### DISCUSSION THREAD

**Discussion by subhashb** [proteanhq/protean#369] [Ideas]
`get` and `find_by` methods in DAOs throw `ObjectNotFoundError` exception on zero results. It would be good to optionally control the exception raising behavior, as needed by the application functionality. Currently, the only way to find an object without raising an exception is with `query.filter(<>).first`.

-----

Behavior in other packages:

SQLAlchemy's `get` returns None if no record was found. Flask SQLAlchemy package has a `get_or_404` method, which is like get() but aborts with 404 if not found instead of returning None.

Django's `get` raises `DoesNotExist` if no objects were found or `MultipleObjectsReturned` if more than one object was found.
There is no `find_by` method in Django.

-----

Options:
1. `get` as-it-is with `find_by` and `find_first_by`
`find_first_by` does not raise exception where as the other two do.
1. `get` as-it-is with `find_by` and `find_by_or_raise`
`find_by` does not raise exception, but `find_by_or_raise`
1. `get` as-it-is with `find_by` and `query.filter(<>).first`
`get` and `find_by` continue to raise exceptions (current behavior).

**Comment by abhirathmahipal** [comment]
For 

Relevant links/discussions we had
- Failing sometimes is better than checking every time.
- Additional test cases to cover the Exception branch (or use #pragma: no cover)


Laying some context to why it'd be useful.

In the try/catch version, we need to test two branches.

```python
try:
    user = user_repo.find_by_email_address('xyz@example.com')
except ObjectNotFoundError:
    return # or handle it by raising another exception

if user.created_at > datetime(2019):
    # do something
    pass

```
--------------


In this version, we only need to test one.
```python

user = user_repo.find_by_email_address('xyz@example.com')

# we need fewer test cases to reach desired coverage here
if user and user.created_at > datetime(2019):
    # do something
    pass
```

**Comment by subhashb** [comment]
Still a real gap: `get()` being fail-loud is right, but there is no clean no-raise lookup (today it is `find(Q)` or reaching into the DAO with `_dao.query.filter().first`). I have written it up as a tracked issue to add `Repository.get_or_none()`: https://github.com/proteanhq/protean/issues/1279. Closing this in favor of the issue.
